Added hint to inner_prod scan

jrmlhermitte opened the pull request:
Some issues with best effort callback. Used a hint in the `relative_inner_product_scan` to help guide the best effort callback.

Sample code to run (toggle `False` to `True` to test between the outer and inner product scans):
(Note this is a copy and paste from documentation)
```python
from bluesky.plan_stubs import abs_set, wait

from bluesky.plans import (relative_outer_product_scan)
from bluesky.plans import relative_inner_product_scan

from bluesky.preprocessors import (run_decorator, stage_decorator,
                                    subs_decorator)
from bluesky.callbacks import LiveTable, LivePlot
from ophyd.sim import det4, motor1, motor2
from bluesky import RunEngine

from bluesky.callbacks.best_effort import BestEffortCallback

from bluesky.utils import ts_msg_hook


bec = BestEffortCallback()

def grid_in_grid(samples):
    """
    Scan a grid around the neighborhood of each sample.

    Parameters
    ----------
    sample : dict
        mapping each sample's name to its (x, y) position
    """

    # In this example we hard-code the hardware and other parameters. For more
    # flexibility, they could instead be parameters to the function.
    detector = det4
    x = motor1
    y = motor2
    x_range = y_range = 0.2
    x_num = y_num = 5

    hints = dict()

    @subs_decorator([
                     LiveTable([detector, x, y]),
                     #LivePlot('motor2'),
                     bec,
                    ])
    def plan():
        for name, position in samples.items():
            # Prepare metadata.
            md = {'sample': name}  #, 'hints': hints}

            # Move to the cetner of the sample position.
            x_pos, y_pos = position
            yield from abs_set(x, x_pos)
            yield from abs_set(y, y_pos)
            yield from wait()

            # Scan a grid around that position.
            if False:
                yield from relative_outer_product_scan([detector],
                                                       x, -x_range, x_range,
                                                       x_num,
                                                       y, -y_range, y_range,
                                                       y_num,
                                                       True, md=md)

            else:
                yield from relative_inner_product_scan([detector],
                                                       x_num,
                                                       x, -x_range, x_range,
                                                       y, -y_range, y_range,
                                                       md=md)

    yield from plan()


# Example usage:

RE = RunEngine({})
RE.msg_hook = ts_msg_hook
RE.subscribe(print)



#bec = BestEffortCallback()
#RE.subscribe(bec)

samples = {'A': (1, 1),
           'B': (1, 2),
           'C': (1, 3),
           'D': (2, 1),
           'E': (2, 2),
           'F': (2, 3)}

RE(grid_in_grid(samples))
```

jrmlhermitte commented:
This is a first pass for #868 and #867.

tacaswell commented on bluesky/plans.py:
```diff
@@ -964,10 +964,18 @@ def relative_inner_product_scan(detectors, num, *args, per_step=None, md=None):
 
+    # update the hints separately for the best effort callback
+    # TODO : consider non-primary?
+    default_hints = {'dimensions': [((motors[0].name,), 'primary')]}
```
The 'primary' here is what data stream these things are in.  It is probably better to ask the motor for it's hints.  The value we care about here is probably `motors[0].user_readback.name` which is most cases is the same as `motors[0].name` but does not have to be.  

We also want to list _all_ of the motors in the dimensions, not just the first one.  This should probably be something like

```python
[(motor.hints.get('fields', ()), 'pirmary') for motor in motors]

```

but I am not sure that that nesting is correct.

jrmlhermitte replied:
sounds good. it should not be a list of motor names but a choice of one for a 1D plot see the modification. not sure how to make it look nicer in terms of list comprehension. @danielballan should have some say as to the general API, if this is the right place to put it

danielballan commented:
To my comment from Slack:

> I think that the plan hints should give the complete story, and it should be up to BEC to decide whether it wants to ignore axes beyond the zeroth one. Someday we might get better high-dimensional viz, and it would be nice if it worked on old data.

As I load my memory of the design discussions around 'hints' back into mental RAM, I think we actually planned for this case (but neglected to implement it!). For a grid scan over two motors, the hint looks like:

```
[(('motor1',), 'primary), ('motor2',), 'primary')]
```

For an inner-product scan with two motors, the hint should look like:

```
[(('motor1', 'motor2'), 'primary')]
```

Each element in the list is an axis of the _scan_. That axis is often composed of just one motor, but as in the case of an inner-product scan, it can be composed of multiple motors moving together. This is why we specify `('motor1',)` is a single-element tuple -- we left room for the more general case.

jrmlhermitte commented:
Ok, here is a proposed change. There are two levels to this:
1. Inner product scan needs to provide a hint that suggests motors are along the same axis, i.e. for two motor names, 'motor1' and 'motor2', changed the assumed `dimension`s from :
```python
[(('motor1',), 'primary), ('motor2',), 'primary')]
```
to
```python
[(('motor1', 'motor2',), 'primary')]
```
That's the attempt I made in `plans.py`
2. Finally, there is an issue where this list is currently flattened and assumed each motor name is a dimension, meaning both these lists would change to:
```python
[(('motor1', 'motor2',), 'primary')]
```
I have modified the code so that we pick only one element of the name tuples in each list (`('motor1', 'motor2')` for example).
Thus we end for both cases with the same flattened list as before, but with the correct number of elements.

TODO (suggestions from Dan) 
- Right now, the first motor is assumed to be the x-axis. It could be interesting to allow the user to choose another motor (or require them to put the motor they want as x-axis first)
- It might be interesting to hold on to the extra motor information, for example, use `'motor2'` for the legend in the `inner_product_scan` case (would not make sense for `outer_product_scan`.

These could be a new PR with issue or handled here.

Thoughts? Suggestions?

tacaswell commented on bluesky/plans.py:
```diff
@@ -964,10 +964,27 @@ def relative_inner_product_scan(detectors, num, *args, per_step=None, md=None):
 
+    # Default should assume the first motor is what to plot
+    # If another motor is desired, override with separate hints
+    if hasattr(motors[0], 'hints') and 'fields' in motors[0].hints:
```
Do we need this conditional?  Could we do this as

```python
for motor in motors:
    fields.extend(getattr(motor, 'hints', {}).get('fields', []))
```

?

jrmlhermitte replied:
yes that's much more elegant thanks! made a few extra changes

danielballan commented:
Elegant! Your helpful comments should rescue us next time we have to remember how this works.

jrmlhermitte commented:
thanks!
